pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/DelayInjectionMetadataStore.java

```diff
@@ -0,0 +1,245 @@
+        if (delay.isPresent()) {
+            return delay.get().thenCompose(__ -> store.put(path, value, expectedVersion, options));
+        }
+        return store.put(path, value, expectedVersion, options);
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(String path, Optional<Long> expectedVersion) {
+        if (maxRandomDelayMills.get() > 0) {
+            return injectRandomDelay(() -> store.delete(path, expectedVersion));
+        }
+
+        Optional<CompletableFuture<Void>> delay = programmedDelays(OperationType.DELETE, path);
+        if (delay.isPresent()) {
+            return delay.get().thenCompose(__ -> store.delete(path, expectedVersion));
+        }
+        return store.delete(path, expectedVersion);
+    }
+
+    @Override
+    public CompletableFuture<Void> deleteRecursive(String path) {
+        if (maxRandomDelayMills.get() > 0) {
+            return injectRandomDelay(() -> store.deleteRecursive(path));
+        }
+
+        Optional<CompletableFuture<Void>> delay = programmedDelays(OperationType.DELETE, path);
+        if (delay.isPresent()) {
+            return delay.get().thenCompose(__ -> store.deleteRecursive(path));
+        }
+        return store.deleteRecursive(path);
+    }
+
+    @Override
+    public void registerListener(Consumer<Notification> listener) {
+        store.registerListener(listener);
+    }
+
+    @Override
+    public <T> MetadataCache<T> getMetadataCache(Class<T> clazz) {
+        return store.getMetadataCache(clazz);
+    }
+
+    @Override
+    public <T> MetadataCache<T> getMetadataCache(TypeReference<T> typeRef) {
+        return store.getMetadataCache(typeRef);
+    }
+
+    @Override
+    public <T> MetadataCache<T> getMetadataCache(MetadataSerde<T> serde) {
+        return store.getMetadataCache(serde);
+    }
+
+    @Override
+    public void registerSessionListener(Consumer<SessionEvent> listener) {
+        store.registerSessionListener(listener);
+        sessionListeners.add(listener);
+    }
+
+    @Override
+    public void close() throws Exception {
+        store.close();
+    }
+
+    public void setMaxRandomDelayMills(int maxRandomDelayMills) {
+        this.maxRandomDelayMills.set(maxRandomDelayMills);
+    }
+
+    public void setMinRandomDelayMills(int minRandomDelayMills) {
+        this.minRandomDelayMills.set(minRandomDelayMills);
+    }
+
+    public void delayConditional(CompletableFuture<Void> wait, BiPredicate<OperationType, String> predicate) {
+        delays.add(new Delay(wait, predicate));
+    }
+
+    public void triggerSessionEvent(SessionEvent event) {
+        sessionListeners.forEach(l -> l.accept(event));
+    }
+
+    private Optional<CompletableFuture<Void>> programmedDelays(OperationType op, String path) {
+        Optional<Delay> failure = delays.stream()
+                .filter(f -> f.predicate.test(op, path))
+                .findFirst();
+
+        failure.ifPresent(delays::remove);
```

Comment:
This operation is not thread safe.
If two threads enter this method they could use the same CompletareFuture leading to unpredictable results.

Reply:
Great catch. 
This will leads to an extra and unexpected delay.